### ISSUE TITLE
Tap dance `on tap` triggered when `on hold` is expected

### ISSUE THREAD

**Issue by LyraelRayne** [vial-kb/vial-qmk#786]
I have the use case that I want `KC_A`/`KC_D` when I `on tap` `A`/`D`, and `KC_LEFT`/`KC_RIGHT` when I `on hold` the respective key (for use with games, where it'd be handy to have an extra 4 action keys, but still be able to move using WASD). 

## What happens
Currently if I set it up this way I get the following behaviour: 
- I press and hold `A` down `KC_LEFT` is held (good!)
- I decide I want to start moving right, I hold `D` then release `A` during the tapping term - `KC_LEFT` is released. (unexpected)
- Release `D` - `KC_D` is tapped once (also unexpected)

## What I expect
- I press and hold `A` down - `KC_LEFT` is held
- I press and hold `D` then release `A` during tapping term - `KC_RIGHT` is held. `KC_LEFT` has not been released yet
- I release `A` - `KC_LEFT` is released

Basically I expect the Tap Dance keys to work at the same time as one another without interfering with the other tap dances and this is not what is happening. This seems to be similar to an old (now fixed) QMK bug https://github.com/qmk/qmk_firmware/issues/527 so perhaps the solution to that might also work for this?
